cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -1340,53 +1346,59 @@ private AppFabricServiceStatus start(final Id.Program id, ProgramType type,
       final ProgramController controller = runtimeInfo.getController();
       final String runId = controller.getRunId().getId();
 
-      controller.addListener(new AbstractListener() {
-
-        @Override
-        public void init(ProgramController.State state, @Nullable Throwable cause) {
-          store.setStart(id, runId, TimeUnit.SECONDS.convert(System.currentTimeMillis(), TimeUnit.MILLISECONDS));
-          if (state == ProgramController.State.COMPLETED) {
-            completed();
+      if (type == ProgramType.MAPREDUCE) {
+        // For MapReduce programs recording of the states in the MDS is done by MapReduceRuntimeService
+        // We only need listener here to record explicit stopping of the MapReduce program
```

Comment:
Why? If the program is stopped by user explicitly, there will be a stop signal send to the program runner so that it can interrupt/terminate whatever is running. The means the program runner does know that user request a stop, hence can record the state in MDS.